### PR TITLE
Rewrite pileup module

### PR DESCRIPTION
**Pull request by alumi**
#### Summary
**BREAKING CHANGES**
Completely rewrote `pileup` module in order to improve its performance and usability.

#### Changes
- Don't pile up uncovered regions
    - Sparsely piled-up alignments
- Extract I/O from pileup algorithm
    - Reader/Writer of pileup format
    - Improved serialization with start/end (^/$) and mapq support
- Move everything in `cljam.pileup.mpileup` to `cljam.pileup`
    - Remove `cljam.pileup.common` which is no longer referred
- Add a benchmarking script

##### Performance comparison
Piling up `chr1:1-247249719` of `large.bam` with a single thread.

###### master
```time: 2301.451698 sec, sd: 40390.179732 sec```

###### feature/sparse-mpileup
```time: 48.248668 sec, sd: 648.517263 µs```

#### Affects
- `cljam.algo.pileup`
- `cljam.io.sam.util.cigar`
- `cljam.io.sam.util.quality`

#### Tests

- `lein check` 🆗
- `lein test :all` 🆗

#### Notes
- It could be even faster if we implement the eager version of pileup like `cljam.algo.depth/depth`
- Upgraded cloverage to `1.0.11`

### CODE REVIEW

**Comment by alumi** [comment]
@totakke Thanks for your helpful review! 🙇 

I added some commits to fix problems you pointed. The `Ġ` was caused by encoding MAPQ 255 into phred33 char without range checks, now fixed with c72c48d.

```bash
bash -c 'diff <(lein run pileup test-resources/bam/small.bam -r chr1:23000088-23000117) <(samtools mpileup test-resources/bam/small.bam -r chr1:23000088-23000117) | head' 2>/dev/null
# no diffs in the region
```

File: src/cljam/algo/pileup.clj [hunk 1]
```diff
@@ -1,17 +1,208 @@
...
+              (map index-cigar)))
+            (pileup-seq s e)
+            (sequence
+             (comp (map resolve-bases)
+                   (map correct-overlaps)
+                   (map (filter-by-base-quality min-base-quality))
+                   (keep (partial ->locus-pile chr)))))))))
+
+(defn align-pileup-seqs
+  "Align multiple pileed-up seqs."
```
Comment (athos) [inline src/cljam/algo/pileup.clj]: s/pileed-up/piled-up/

File: src/cljam/algo/pileup.clj [hunk 2]
```diff
@@ -1,17 +1,208 @@
...
+                                           :filtered-out
+                                           :duplicated
+                                           :supplementary
+                                           :secondary}) flag))
+           (or (not (flag/multiple? flag)) (flag/properly-aligned? flag))
+           (not-empty (.cigar aln))
+           (<= min-mapq (.mapq aln))))))
+
+(defn resolve-base
+  "Find a piled-up base and an indel from a alignment."
```
Comment (athos) [inline src/cljam/algo/pileup.clj]: "an alignment" or "the alignment"

**Comment by alumi** [comment]
@athos Thanks for your review!! 🙇
I added `alignment` field of `PileupBase` just for a use of the direct accessor `.alignment`.
So I change it to an extra entry and add `qname` field instead.

**Comment by totakke** [comment]
Thank you!

**Comment by athos** [comment]
Thank you for tackling a tough work, @alumi!

**Pull request merged**